[DOCS] Rework semantic search main page (#112452)
elastic/elasticsearch: docs/reference/search/search-your-data/semantic-search.asciidoc

You can also deploy NLP in {es} manually, without using an {infer} endpoint.
This is the most complex and labor intensive workflow for performing semantic search in the {stack}.
You need to select an NLP model from the {ml-docs}/ml-nlp-model-ref.html#ml-nlp-model-ref-text-embedding[list of supported dense and sparse vector models], deploy it using the Eland client, create an index mapping, and set up a suitable ingest pipeline to start ingesting and querying data.
the <<semantic-search-deployed-nlp-model, Semantic search with a model deployed in {es}>> tutorial shows you the process end-to-end.

[discrete]
[[using-query]]
=== Using the right query

Crafting the right query is crucial for semantic search.
Which query you use and which field you target in your queries depends on your chosen workflow.
If you're using the `semantic_text` workflow it's quite simple.
If not, it depends on which type of embeddings you're working with.

[cols="3*", options="header"]
|=======================================================================================================================================================================================================

> | Field type to query                    | Query to use                                      | Notes                                                                                                                                                             
> | <<semantic-text,`semantic_text`>> .    | <<query-dsl-semantic-query,`semantic`>> .         | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     
> | <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
> | <<dense-vector,`dense_vector`>>        | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
> |=======================================================================================================================================================================================================

Author: ```suggestion
| Field type to query                    | Query to use                                      | Notes                                                                                                                                                             
| <<semantic-text,`semantic_text`>>      | <<query-dsl-semantic-query,`semantic`>>           | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     
| <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
| <<dense-vector,`dense_vector`>>        | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
|=======================================================================================================================================================================================================
```

* use the `semantic_text` workflow (recommended)
* use the {infer} API workflow
* deploy models directly in {es}

Refer to <<using-nlp-models,this section>> to choose your workflow.

You can also store your own embeddings in {es} as vectors.
Refer to <<using-query,this section>> for guidance on which query type to use for semantic search.

At query time, {es} can use the same NLP model to convert a query into embeddings, enabling you to find documents with similar text embeddings.

[discrete]
[[using-nlp-models]]
=== Choose a semantic search workflow

[discrete]
==== The `semantic_text` workflow

The simplest way to use NLP models in the {stack} is through the <<semantic-search-semantic-text, `semantic_text` workflow>>.
We recommend using this approach because it abstracts away a lot of manual work.
All you need to do is create an {infer} endpoint and an index mapping to start ingesting, embedding, and querying data.
There is no need to define model-related settings and parameters, or to create {infer} ingest pipelines.
Refer to the <<put-inference-api, Create an {infer} endpoint API>> documentation for a list of supported services.

> The <<semantic-search-semantic-text, Semantic search with `semantic_text`>> tutorial shows you the process end-to-end.

Reviewer: ```suggestion

The <<semantic-search-semantic-text, Semantic search with `semantic_text`>> tutorial shows you the process end-to-end.
```

[discrete]
==== The {infer} API workflow

The <<semantic-search-inference, {infer} API workflow>> is more complex but offers greater control over the {infer} endpoint configuration.
You need to create an {infer} endpoint, provide various model-related settings and parameters, define an index mapping, and set up an {infer} ingest pipeline with the appropriate settings.

> The <<semantic-search-inference, Semantic search with the {infer} API>> tutorial shows you the process end-to-end.

Reviewer: ```suggestion

The <<semantic-search-inference, Semantic search with the {infer} API>> tutorial shows you the process end-to-end.
```

[discrete]
==== Model deployment workflow

You can also deploy NLP in {es} manually, without using an {infer} endpoint.
This is the most complex and labor intensive workflow for performing semantic search in the {stack}.
You need to select an NLP model from the {ml-docs}/ml-nlp-model-ref.html#ml-nlp-model-ref-text-embedding[list of supported dense and sparse vector models], deploy it using the Eland client, create an index mapping, and set up a suitable ingest pipeline to start ingesting and querying data.

> the <<semantic-search-deployed-nlp-model, Semantic search with a model deployed in {es}>> tutorial shows you the process end-to-end.

Reviewer: ```suggestion

The <<semantic-search-deployed-nlp-model, Semantic search with a model deployed in {es}>> tutorial shows you the process end-to-end.
```

[discrete]
[[using-query]]
=== Using the right query

Crafting the right query is crucial for semantic search.
Which query you use and which field you target in your queries depends on your chosen workflow.
If you're using the `semantic_text` workflow it's quite simple.
If not, it depends on which type of embeddings you're working with.

> [cols="3*", options="header"]

Reviewer: nit: The table columns are a bit disorienting, wonder if could give a bit more width to the first two?

|=======================================================================================================================================================================================================
| Field type to query                    | Query to use                                      | Notes                                                                                                                                                             
| <<semantic-text,`semantic_text`>>      | <<query-dsl-semantic-query,`semantic`>>           | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     

> | <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.

Reviewer: ```suggestion
| <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse_vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
```

the underscore might help with column lengths too?

| <<dense-vector,`dense_vector`>>        | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
|=======================================================================================================================================================================================================

> If you want {es} to generate embeddings for you both index and query time, use the `semantic_text` field and the `semantic` query.

Reviewer: ```suggestion
If you want {es} to generate embeddings at both index and query time, use the `semantic_text` field and the `semantic` query.
```

Using an NLP model enables you to extract text embeddings out of text.
Embeddings are vectors that provide a numeric representation of a text.
Pieces of content with similar meaning have similar representations.

You have several options for using NLP models in the {stack}:

* use the `semantic_text` workflow (recommended)
* use the {infer} API workflow
* deploy models directly in {es}

Refer to <<using-nlp-models,this section>> to choose your workflow.

You can also store your own embeddings in {es} as vectors.
Refer to <<using-query,this section>> for guidance on which query type to use for semantic search.

At query time, {es} can use the same NLP model to convert a query into embeddings, enabling you to find documents with similar text embeddings.

[discrete]
[[using-nlp-models]]
=== Choose a semantic search workflow

[discrete]

> ==== The `semantic_text` workflow

Reviewer 2: ```suggestion
==== `semantic_text` workflow
```

The simplest way to use NLP models in the {stack} is through the <<semantic-search-semantic-text, `semantic_text` workflow>>.
We recommend using this approach because it abstracts away a lot of manual work.
All you need to do is create an {infer} endpoint and an index mapping to start ingesting, embedding, and querying data.
There is no need to define model-related settings and parameters, or to create {infer} ingest pipelines.
Refer to the <<put-inference-api, Create an {infer} endpoint API>> documentation for a list of supported services.

The <<semantic-search-semantic-text, Semantic search with `semantic_text`>> tutorial shows you the process end-to-end.

[discrete]

> ==== The {infer} API workflow

Reviewer 2: ```suggestion
==== {infer} API workflow
```

[discrete]
[[using-query]]
=== Using the right query

Crafting the right query is crucial for semantic search.
Which query you use and which field you target in your queries depends on your chosen workflow.
If you're using the `semantic_text` workflow it's quite simple.
If not, it depends on which type of embeddings you're working with.

[cols="30%, 30%, 40%", options="header"]
|=======================================================================================================================================================================================================
| Field type to query                    | Query to use                                      | Notes                                                                                                                                                             
| <<semantic-text,`semantic_text`>>      | <<query-dsl-semantic-query,`semantic`>>           | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                               

> | <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse_vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. Provide embeddings at index time.

Reviewer 2: ```suggestion
| <<sparse-vector,`sparse_vector`>>      | <<query-dsl-sparse-vector-query,`sparse_vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You must provide embeddings at index time.
```

> | <<dense-vector,`dense_vector`>>        | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. Provide embeddings at index time.

Reviewer 2: ```suggestion
| <<dense-vector,`dense_vector`>>        | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. You must provide embeddings at index time.
```

|=======================================================================================================================================================================================================

If you want {es} to generate embeddings at both index and query time, use the `semantic_text` field and the `semantic` query.

> If you want to bring your own embeddings, store them in {es} and use the `sparse_vector` or `dense_vector` field type and the associated query depending on the NLP model you used for generating the embeddings.

Reviewer 2: ```suggestion
If you want to bring your own embeddings, use the `sparse_vector` or `dense_vector` field type and the associated query depending on the NLP model you used to generate the embeddings.
```

IMPORTANT: For the easiest way to perform semantic search in the {stack}, refer to the <<semantic-search-semantic-text, `semantic_text`>> end-to-end tutorial.

[discrete]
[[semantic-search-read-more]]
=== Read more

* Tutorials:
** <<semantic-search-semantic-text, Semantic search with `semantic_text`>>
** <<semantic-search-inference, Semantic search with the {infer} API>>

> ** <<semantic-search-elser,Semantic search with ELSER>> using the {infer} workflow

Reviewer 2: ```suggestion
** <<semantic-search-elser,Semantic search with ELSER>> using the model deployment workflow
```

** <<semantic-search-deployed-nlp-model, Semantic search with a model deployed in {es}>>
** {ml-docs}/ml-nlp-text-emb-vector-search-example.html[Semantic search with the msmarco-MiniLM-L-12-v3 sentence-transformer model]
* Interactive examples:

> ** The https://github.com/elastic/elasticsearch-labs[`elasticsearch-labs`] repo contains a number of interactive semantic search examples in the form of executable Python notebooks, using the {es} Python client

Reviewer 2: Can we link to some specific examples here?

Semantic search with ELSER using the model deployment workflow: https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/03-ELSER.ipynb

Semantic search with `semantic_text`: https://github.com/elastic/elasticsearch-labs/blob/main/notebooks/search/09-semantic-text.ipynb

> * Blogs:

Reviewer 2: Can we link to the [`semantic_text` blog post](https://www.elastic.co/search-labs/blog/semantic-search-simplified-semantic-text) in this section?

Pieces of content with similar meaning have similar representations.
NLP models can be used in the {stack} various ways, you can:

* use the <<semantic-search-semantic-text, `semantic_text` workflow>> (recommended)
* use the <<semantic-search-inference, {infer} API workflow>>

> * deploy models directly in {es}

Author: I'd highlight the recommended option. If we label all three they seem like semantic search is for newcomers while the other two are for more advanced users but AFAIK it's not our intention. WDYT?

There is no need to define model-related settings and parameters, or to create {infer} ingest pipelines.
Refer to the <<put-inference-api, Create an {infer} endpoint API>> documentation for a list of supported services.

The <<semantic-search-inference, {infer} API workflow>> more complex but offers greater control over the {infer} endpoint configuration.
You need to create an {infer} endpoint, provide various model-related settings and parameters, define an index mapping, and set up an {infer} ingest pipeline with the appropriate settings.

You can also deploy NLP models directly in {es}, which is the most complex way to perform semantic search in the {stack}.
You need to select an NLP model from the {ml-docs}/ml-nlp-model-ref.html#ml-nlp-model-ref-text-embedding[list of supported dense and sparse vector models], deploy it using the Eland client, create an index mapping, and set up a suitable ingest pipeline to start ingesting and querying data.

[discrete]
[[using-query]]
=== Using the right query

Crafting the right query is crucial for semantic search.

> The query type you should use depends first on whether you are using the recommended `semantic_text` workflow.
> If not, it depends on the vector type in which your embeddings are stored.

Reviewer: ```suggestion
Which query you use and which field you target in your queries depends on your chosen workflow.
If you're using the `semantic_text` workflow it's quite simple.
If not, it depends on which type of embeddings you're working with.
```

[cols="3*", options="header"]
|=======================================================================================================================================================================================================

> | Field to query                    | Query to use                                      | Notes                                                                                                                                                             

Reviewer: ```suggestion
| Field type to query                    | Query to use                                      | Notes                                                                                                                                                             
```

| <<semantic-text,`semantic_text`>> | <<query-dsl-semantic-query,`semantic`>> .         | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     
| <<sparse-vector,`sparse_vector`>> | <<query-dsl-sparse-vector-query,`sparse vector`>> | The `sparse_vector` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
| <<dense-vector,`dense_vector`>>   | <<query-dsl-knn-query,`knn`>>                     | The `knn` query can generate query embeddings for you, but you can also provide your own. You are expected to provide embeddings at index time.
|=======================================================================================================================================================================================================

If you want {es} to generate embeddings for you both index and query time, use the `semantic_text` field and the `semantic` query.

> If you want to bring your own embeddings, store them in {es} and use the `sparse_vector` or `dense_vector` field type and the associated query depending on the NLP model you used for generating the embeddings.

Reviewer: Would like to see mention of the BYOE use case earlier in this doc somewhere too.

IMPORTANT: For the easiest way to perform semantic search in the {stack}, refer to the <<semantic-search-semantic-text, `semantic_text`>> end-to-end tutorial.

[discrete]
[[semantic-search-read-more]]
=== Read more

> * Tutorials:

Author: Great idea. However, I'd keep this section at the end of the page to have all the related materials in one place for ease of reference.

You need to select an NLP model from the {ml-docs}/ml-nlp-model-ref.html#ml-nlp-model-ref-text-embedding[list of supported dense and sparse vector models], deploy it using the Eland client, create an index mapping, and set up a suitable ingest pipeline to start ingesting and querying data.

[discrete]
[[using-query]]
=== Using the right query

Crafting the right query is crucial for semantic search.
Which query you use and which field you target in your queries depends on your chosen workflow.
If you're using the `semantic_text` workflow it's quite simple.
If not, it depends on which type of embeddings you're working with.

[cols="3*", options="header"]
|=======================================================================================================================================================================================================
| Field type to query                    | Query to use                                      | Notes                                                                                                                                                             

> | <<semantic-text,`semantic_text`>> | <<query-dsl-semantic-query,`semantic`>> .         | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     

Reviewer: ```suggestion
| <<semantic-text,`semantic_text`>> | <<query-dsl-semantic-query,`semantic`>>          | The `semantic_text` field handles generating embeddings for you at index time and query time.                                                                     
```
